src/core/Akka.Streams/Dsl/RepeatPrevious.cs

```diff
@@ -0,0 +1,105 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="RepeatPrevious.cs" company="Akka.NET Project">
+//     Copyright (C) 2013-2022 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using Akka.Streams.Stage;
+using Akka.Util;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// Delegate used to customize replacement functions.
+    ///
+    /// This is used for things like calling <see cref="IDisposable.Dispose"/> on the previous element.
+    /// when a new value is emitted.
+    /// </summary>
+    /// <typeparam name="T">The type of element handled by the <see cref="RepeatPrevious{T}"/></typeparam>
+    public delegate void SwapPrevious<T>(T previousValue, T newValue);
+
+    /// <summary>
+    /// Repeats the previous element from upstream until it's replaced by a new value.
+    ///
+    /// This is designed to allow fan-in stages where output from one of the sources is intermittent / infrequent
+    /// and users just want the previous value to be reused.
+    /// </summary>
+    /// <typeparam name="T">The output type.</typeparam>
+    public sealed class RepeatPrevious<T> : GraphStage<FlowShape<T, T>>
+    {
+        private readonly Inlet<T> _in = new Inlet<T>("RepeatPrevious.in");
+        private readonly Outlet<T> _out = new Outlet<T>("RepeatPrevious.out");
+
+        public override FlowShape<T, T> Shape => new FlowShape<T, T>(_in, _out);
+        private readonly SwapPrevious<T> _swapPrevious;
+
+        /// <summary>
+        /// Do nothing by default
+        /// </summary>
+        private static readonly SwapPrevious<T> DefaultSwap = (value, newValue) => { };
+
+        public RepeatPrevious() : this(DefaultSwap)
+        {
+        }
+
+        public RepeatPrevious(SwapPrevious<T> swapPrevious)
+        {
+            _swapPrevious = swapPrevious;
+        }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) =>
+            new Logic(this, _swapPrevious);
+
+        private sealed class Logic : InAndOutGraphStageLogic
```

Comment:
I will say that overall this reminds me of a version of `Expand` with a bit more sane of an API. I'm not sure if that means we can de-dupe any logic between the two however (Expand is... weird to me.)

Reply:
I had no idea about `Expand` TBH - part of why I did what I did in the documentation with this stage is that I think our overview of built-in stages is pretty weak and not easy to grok.